src/storage/mvcc/reader.rs

```diff
@@ -401,9 +413,140 @@ impl MvccReader {
+            }
+
+            let mut write = {
+                let (commit_ts, key) = {
+                    let w_cur = self.write_cursor.as_ref().unwrap();
+                    last_handled_key = Some(w_cur.key().to_vec());
+                    let w_key = Key::from_encoded(w_cur.key().to_vec());
+                    (w_key.decode_ts()?, w_key.truncate_ts()?)
+                };
+
+                // reach neighbour user key or can't see this version.
+                if ts < commit_ts || &key != user_key {
+                    assert!(&key <= user_key);
+                    return self.get_value(user_key, lastest_version.0, lastest_version.1);
+                }
+                self.statistics.write.processed += 1;
+                Write::parse(self.write_cursor.as_ref().unwrap().value())?
+            };
+
+            match write.write_type {
+                WriteType::Put => {
+                    if write.short_value.is_some() {
+                        if self.key_only {
+                            lastest_version = (None, Some(vec![]));
+                        } else {
+                            lastest_version = (None, write.short_value.take());
+                        }
+                    } else {
+                        lastest_version = (Some(write.start_ts), None);
+                    }
+                }
+                WriteType::Delete => lastest_version = (None, None),
+                WriteType::Lock | WriteType::Rollback => {}
+            }
+            self.write_cursor
+                .as_mut()
+                .unwrap()
+                .prev(&mut self.statistics.write);
+        }
+
+        // After several prev, we still not get the latest version for the specified ts,
+        // use seek to locate the latest version.
+        let key = user_key.append_ts(ts);
+        let valid = self.write_cursor
+            .as_mut()
+            .unwrap()
+            .internal_seek(&key, &mut self.statistics.write)?;
+        assert!(valid);
+        loop {
+            let mut write = {
+                // If we reach the last handled key, it means we have checked all versions
+                // for this user key.
+                if self.write_cursor.as_ref().unwrap().key()
+                    >= last_handled_key.as_ref().unwrap().as_slice()
+                {
+                    return self.get_value(user_key, lastest_version.0, lastest_version.1);
+                }
+
+                let w_cur = self.write_cursor.as_ref().unwrap();
+                let w_key = Key::from_encoded(w_cur.key().to_vec());
+                let commit_ts = w_key.decode_ts()?;
+                assert!(commit_ts <= ts);
+                let key = w_key.truncate_ts()?;
+                assert_eq!(&key, user_key);
+                self.statistics.write.processed += 1;
+                Write::parse(w_cur.value())?
+            };
+
+            match write.write_type {
```

Comment:
There are several similar blocks for `match write.write_type`, could we define a function to remove the repetition? Maybe we could do it in another PR.

Reply:
I think this will be done in refine Cursor and MvccRedeader task.